### DISCUSSION TITLE
garden config options additional-host-entry and allow-host-access

### DISCUSSION THREAD

**Discussion by AnianZ** [concourse/concourse#6549] [Help & Support]
I'm currently using a `gdn-config.ini` looking like this:
```
[server]
allow-host-access = true
additional-host-entry = 178.3.23.162 something.local
``` 

My use case for this is that I need to access a host from some jobs which is only reachable via a WireGuard VPN from the worker host. I did not find a good option to make that route available to the containers in garden, so I bound the remote port to a local port using `socat` and allowed the containers to access the network on the host and hid the hosts IP behind an additional host entry to allow multiple workers to work smoothly. So the IP in the config is different on every worker.

If the default runtime becomes containerd how does that translate? Are there similar options with containerd or a better way to do this? I assume that my solution for garden is far from optimal anyway.

**Comment by IvanChalukov** [comment]
We’re currently using the same approach as @AnianZ described in our deployments that still rely on the Guardian runtime. Since we’re planning to migrate to the containerd runtime — especially with it becoming the default soon — we’re exploring how to replicate this behavior there.

@taylorsilva do you know if there’s a way to achieve this with containerd? I looked into the codebase but couldn’t find an equivalent configuration for injecting additional `/etc/hosts` entries like we can with Guardian.

As a potential path forward, I was wondering if it might be feasible to introduce a new configuration parameter, something like `CONCOURSE_CONTAINERD_ADDITIONAL_HOST_ENTRIES`, which could be handled during container creation (e.g. when starting a new task [here](https://github.com/concourse/concourse/blob/master/worker/runtime/backend.go#L362)).

Not sure how realistic that is — just thinking out loud and hoping to explore potential solutions.

**Reply by taylorsilva** [reply]
Need to investigate, but wondering if we can do something similar to what we do for `/etc/resolv.conf`? That was the first thought I had and would look into first. Not sure if containerd has anything built-in that would help with that.

**Reply by ceco556** [reply]
Hi @taylorsilva, unfortunately containerd doesn't have a built-in option to handle entries for `/etc/hosts`. Ivan's idea is good but I am not sure there if there is something we can do in the Concourse codebase to handle this if it doesn't come from the containerd as an option. 

It might be better to raise this topic in the containerd discussions? What do you think?

**Reply by Kump3r** [reply]
Not an expert here, but based on the fact how, overwriting the dns servers in resolve.conf works [`generateResolvConfContents`](https://github.com/concourse/concourse/blob/61bb161329421e7bb2103f6e865ed1b8a54ae499/worker/runtime/cni_network.go#L231-L232). I think [etcHosts](https://github.com/concourse/concourse/blob/61bb161329421e7bb2103f6e865ed1b8a54ae499/worker/runtime/cni_network.go#L222) in the cniNetwork would be also a feasible place to create additional hosts from a parameter passed from the ENV variables, to ensure consistency with `guardian` runtime wdyt? As Taylor mentioned similar to https://github.com/concourse/concourse/pull/5445

**Reply by taylorsilva** [reply]
I suspect there is a way to do this with containerd as-is. Both k8s and Docker allow adding extra entries to `/etc/hosts`:

- https://kubernetes.io/docs/tasks/network/customize-hosts-file-for-pods/
- https://docs.docker.com/reference/cli/docker/container/run/#add-host

And both use containerd. Just need to do some digging or playing around with the code.

What Kalin pointed to is a good spot to start. I'd start there as well with trying to POC something.

**Reply by Kump3r** [reply]
For reslov.conf, we based the implementation based on Guardian, so I assume a similar should be done here:
[Reslov.conf Guardian](https://github.com/cloudfoundry/guardian/blob/main/kawasaki/dns/resolv_compiler.go) -> [Resolv.conf Concourse](https://github.com/concourse/concourse/blob/master/worker/runtime/resolvconf_parser.go)
[Hosts Guardian](https://github.com/cloudfoundry/guardian/blob/main/kawasaki/dns/hosts_file_compiler.go) -> ...

Edit: I do not think we actually need this much detail after further looking into this with @ceco556 we drafted a small POC (~~still to be validated~~), with which we aim to introduce the change. Hardcoding some values directly with a small POC for cycle in the SetupMounts func worked, so now we are figuring where exactly to populate all the commands, so the worker process is started with the props.
Tested it and it seems to work.
<img width="1040" height="257" alt="Screenshot 2025-07-23 at 17 38 05" src="https://github.com/user-attachments/assets/55b8e189-c3da-4a6b-bc47-cff6c59823d8" />

TLDR:
~~I will open a DRAFT PR~~, so you can join with ideas, tests and documentation. I will also try to prepare PR's for bosh/helm charts
https://github.com/concourse/concourse/pull/9238

**Reply by Kump3r** [reply]
Created `bosh`, `chart` and `docs` PR's for it. @taylorsilva can you confirm if the overall style is okay based on a short screening, so we can contribute some tests. Or if you have some scenarios that would like covered, feel free to join and contribute, I added you to my fork, so you could jump in. Thanks to @ceco556 for collaborating with me and for his hardcoded POC approach.
